### PR TITLE
Ensure relay chain block number strictly increases

### PR DESCRIPTION
**Pull request by bkchr**
This is a safeguard for when async backing is enabled on the relay chain and multiple parachain
blocks per relay chain block are legal. We will need to later change this to support then multiple
parachain blocks per relay chain block.

### CODE REVIEW

**Comment by xlc** [comment]
This means we cannot easily swap the relaychain of a testnet parachain, which is a very useful feature to maintain a long term testnet.

**Comment by bkchr** [comment]
Stuff like the latest message head already prevent this (it may currently works for you because you don't have send any messages on your test chain).

However, I can make this configurable from the beginning.

**Comment by xlc** [comment]
The latest message head can be fixed by setting storages on relaychain with sudo. We can't just bump block number on relaychain (or can we?)

**Review by rphmeier — APPROVED** [review]

**Comment by bkchr** [comment]
> The latest message head can be fixed by setting storages on relaychain with sudo. We tried this with last rococo reset. We can't just bump block number on relaychain (or can we?)

No, but if I make it configurable, you can ignore this test on your test parachain.

**Comment by bkchr** [comment]
> The latest message head can be fixed by setting storages on relaychain with sudo. We tried this with last rococo reset. We can't just bump block number on relaychain (or can we?)

No, but if I make it configurable, you can ignore this test on your test parachain.

**Review by gilescope — APPROVED** [review]

**Comment by xlc** [comment]
> > The latest message head can be fixed by setting storages on relaychain with sudo. We tried this with last rococo reset. We can't just bump block number on relaychain (or can we?)
> 
> No, but if I make it configurable, you can ignore this test on your test parachain.

That works. Can we make it configurable? Should I make a new issue?

**Comment by bkchr** [comment]
I will push the changes to this pr before merging it.

**Comment by bkchr** [comment]
I have made it configurable. However, my naming is really shitty. @xlc or @rphmeier someone of you may have a better idea for the naming?

**Review by rphmeier — APPROVED** [review]

**Comment by xlc** [comment]
Thanks. The name looks fine to me.

**Review by dvdplm — APPROVED** [review]

File: pallets/parachain-system/src/lib.rs [hunk 1]
```diff
@@ -474,6 +529,11 @@ pub mod pallet {
 	#[pallet::storage]
 	pub(super) type DidSetValidationCode<T: Config> = StorageValue<_, bool, ValueQuery>;
 
+	/// The relay chain block number associated to the last parachain block.
```
Comment (andresilva) [inline pallets/parachain-system/src/lib.rs:532]:
```suggestion
	/// The relay chain block number associated with the last parachain block.
```

File: pallets/parachain-system/src/tests.rs [hunk 2]
```diff
@@ -107,6 +107,7 @@ impl Config for Test {
 	type ReservedDmpWeight = ReservedDmpWeight;
 	type XcmpMessageHandler = SaveIntoThreadLocal;
 	type ReservedXcmpWeight = ReservedXcmpWeight;
+	type CheckAssociatedRelayNumber = AnyRelayNumber;
```
Comment (andresilva) [inline pallets/parachain-system/src/tests.rs:110]: Shouldn't this be `RelayNumberStrictlyIncreases` for the test below to work?

File: pallets/parachain-system/src/lib.rs [hunk 3]
```diff
@@ -88,6 +88,51 @@ pub use relay_state_snapshot::{MessagingStateSnapshot, RelayChainStateProof};
 
 pub use pallet::*;
 
+/// Something that can check the associated relay block number.
+///
+/// Each Parachain block is build in the context of a relay chain block. This relay chain block
+/// has a block number associated this trait is about. With async backing it is legal to build
```
Comment (andresilva) [inline pallets/parachain-system/src/lib.rs:94]:
```suggestion
/// Each Parachain block is built in the context of a relay chain block, this trait allows us
/// to validate the given relay chain block number. With async backing it is legal to build
```

**Review by andresilva — COMMENTED** [review]
lgtm just need to fix the test

File: pallets/parachain-system/src/lib.rs (hunk 3 above)
Comment (andresilva) [reply pallets/parachain-system/src/lib.rs:94]: Although I think this part is a bit redundant.

**Pull request merged**